pyFAI/ext/CSC_common.pxi

```diff
@@ -0,0 +1,308 @@
+            data_t[::1] std = numpy.empty(self.output_size, dtype=data_d)
+            data_t[::1] sem = numpy.empty(self.output_size, dtype=data_d)
+            data_t[::1] cvariance, cdark, cflat, cpolarization, csolidangle, cabsorption
+            bint do_azimuthal_variance = error_model is ErrorModel.AZIMUTHAL
+            bint do_variance, is_valid, do_dark, do_flat, do_polarization, check_dummy, do_solidangle, do_absorption
+            preproc_t value
+            acc_t delta1, delta2, b, omega_A, omega_B, omega2_A, omega2_B
+
+
+        
+        assert weights.size == self.input_size, "weights size"
+        empty = dummy if dummy is not None else self.empty
+
+
+        if (dummy is not None) and (delta_dummy is not None):
+            check_dummy = True
+            cdummy = <data_t> float(dummy)
+            cddummy = <data_t> float(delta_dummy)
+        elif (dummy is not None):
+            check_dummy = True
+            cdummy = <data_t> float(dummy)
+            cddummy = 0.0
+        else:
+            check_dummy = False
+            cdummy = <data_t> float(empty)
+            cddummy = 0.0
+    
+        if variance is not None:
+            assert variance.size == self.input_size, "variance size"
+            do_variance = True
+            cvariance = numpy.ascontiguousarray(variance.ravel(), dtype=data_d)
+        else:
+            do_variance = error_model is not ErrorModel.NO
+    
+        if dark is not None:
+            do_dark = True
+            assert dark.size == self.input_size, "dark current array size"
+            cdark = numpy.ascontiguousarray(dark.ravel(), dtype=data_d)
+        else:
+            do_dark = False
+        if flat is not None:
+            do_flat = True
+            assert flat.size == self.input_size, "flat-field array size"
+            cflat = numpy.ascontiguousarray(flat.ravel(), dtype=data_d)
+        else:
+            do_flat = False
+        if polarization is not None:
+            do_polarization = True
+            assert polarization.size == self.input_size, "polarization array size"
+            cpolarization = numpy.ascontiguousarray(polarization.ravel(), dtype=data_d)
+        else:
+            do_polarization = False
+        if solidangle is not None:
+            do_solidangle = True
+            assert solidangle.size == self.input_size, "Solid angle array size"
+            csolidangle = numpy.ascontiguousarray(solidangle.ravel(), dtype=data_d)
+        else:
+            do_solidangle = False
+        if absorption is not None:
+            do_absorption = True
+            assert absorption.size == self.input_size, "absorption array size"
+            cabsorption = numpy.ascontiguousarray(absorption.ravel(), dtype=data_d)
+        else:
+            do_absorption = False
+    
+        with nogil:
+            #first loop for pixel in input image
+            for idx in range(self._indptr.shape[0]-1):  
```

Comment:
For 1D outputs that do not fit into L1 cache, it might make sense to loop over tiles on the input image.

Reply:
For 1D output, it does not need to fit completely into the cache since one pixel contributes to adjacent bins. This is probably the reason why it is so fast.

2d output will probably not fit into the L1 cache and bins there can be non contiguous...